wpeview/src/main/cpp/Runtime/WKWebView.cpp

```diff
@@ -43,6 +43,101 @@ void handleCommitBuffer(void* context, WPEAndroidBuffer* buffer, int fenceID)
 
 const int httpErrorsStart = 400;
 
+class SslErrorHandler final {
+public:
+    static SslErrorHandler* createHandler(
+        WebKitWebView* webView, GTlsCertificate* certificate, const char* failingURI, gchar** certificatePEM) noexcept
+    {
+        if (!webView || !certificate || !failingURI || !*failingURI || !certificatePEM) {
+            return nullptr;
+        }
+
+        gchar* host = nullptr;
+        // NOLINTBEGIN(clang-analyzer-optin.core.EnumCastOutOfRange)
+        GUri* uri
+            = g_uri_parse(failingURI, static_cast<GUriFlags>(G_URI_FLAGS_PARSE_RELAXED | G_URI_FLAGS_ENCODED), nullptr);
+        // NOLINTEND(clang-analyzer-optin.core.EnumCastOutOfRange)
+        if (uri) {
+            const char* str = g_uri_get_host(uri);
+            if (str) {
+                host = g_strdup(str);
+            }
+            g_uri_unref(uri);
+        }
+        if (!host) {
+            return nullptr;
+        }
+
+        *certificatePEM = nullptr;
+        g_object_get(certificate, "certificate-pem", certificatePEM, nullptr);
+        if (!*certificatePEM) {
+            g_free(host);
+            return nullptr;
+        }
+
+        return new SslErrorHandler(webView, certificate, g_strdup(failingURI), host);
+    }
+
+    ~SslErrorHandler()
+    {
+        handlingFinished();
+        g_weak_ref_clear(&m_webViewWeakRef);
+    }
+
+    SslErrorHandler(SslErrorHandler&&) = delete;
+    SslErrorHandler& operator=(SslErrorHandler&&) = delete;
+    SslErrorHandler(const SslErrorHandler&) = delete;
+    SslErrorHandler& operator=(const SslErrorHandler&) = delete;
+
+    void acceptCertificate() noexcept
+    {
+        auto* webView = static_cast<WebKitWebView*>(g_weak_ref_get(&m_webViewWeakRef));
+        if (webView) {
+            auto* networkSession = webkit_web_view_get_network_session(webView);
+            if (networkSession) {
+                webkit_network_session_allow_tls_certificate_for_host(networkSession, m_certificate, m_host);
+                g_object_unref(networkSession);
+
+                webkit_web_view_load_uri(webView, m_failingURI);
+            }
+            g_object_unref(webView);
+        }
+        handlingFinished();
+    }
+
+    void rejectCertificate() noexcept { handlingFinished(); }
+
+private:
+    GWeakRef m_webViewWeakRef = {};
+    GTlsCertificate* m_certificate = nullptr;
+    gchar* m_failingURI = nullptr;
+    gchar* m_host = nullptr;
+
+    SslErrorHandler(WebKitWebView* webView, GTlsCertificate* certificate, gchar* failingURI, gchar* host) noexcept
+        : m_failingURI(failingURI)
+        , m_host(host)
+    {
+        g_weak_ref_init(&m_webViewWeakRef, webView);
+        m_certificate = g_object_ref(certificate);
+    }
+
+    void handlingFinished() noexcept
+    {
+        g_weak_ref_set(&m_webViewWeakRef, nullptr);
+
+        if (m_certificate) {
+            g_object_unref(m_certificate);
+            m_certificate = nullptr;
+        }
```

Comment:
We have a handy macro that does this:

```suggestion
        g_clear_object(&m_certificate);
```